docs/sql-migration-guide-upgrade.md

```diff
@@ -130,6 +130,8 @@ license: |
 
   - Since Spark 3.0, Spark will cast `String` to `Date/TimeStamp` in binary comparisons with dates/timestamps. The previous behaviour of casting `Date/Timestamp` to `String` can be restored by setting `spark.sql.legacy.typeCoercion.datetimeToString` to `true`.
 
+  - Since Spark 3.0, if files or subdirectories disappear during recursive directory listing (i.e. they appear in an intermediate listing but then cannot be read or listed during later phases of the recursive directory listing, due to either concurrent file deletions or object store consistency issues) then the listing will fail with an exception unless `spark.sql.files.ignoreMissingFiles` is `true` (default `false`). In previous versions, these missing files or subdirectories would be ignored. Note that this change of behavior only applies during initial table file listing (or during `REFRESH TABLE`), not during query execution: the net change is that `spark.sql.files.ignoreMissingFiles` is now obeyed during table file listing / query planning, not only at query execution time.
```

Comment:
What I am not sure is that .. at least .. I know one user group dependent on this behaviour and presumably (by  #15153), another user group will be affected - it's quite a breaking change ..

Do you think it makes sense:

```diff
     .doc("Whether to ignore missing files. If true, the Spark jobs will continue to run when " +
       "encountering missing files and the contents that have been read will still be returned.")
     .booleanConf
-    .createWithDefault(false)
+    .createOptional
```

and make it `true` by default here and `false` in existing code base?

I was thinking about documenting:
- if it's unset, it follows old behaviours, execution time it's false by default but true when listing up at driver side.
- If we explicitly set true, ignore at both execution and driver
- If we explicitly set false, fail fast at both sides.

Reply:
Out of curiosity, can you give an example of a user workload which is dependent on ignoring files that go listing between the initial listing and a recursive listing / stat? I understand the case of wanting to ignore completely-missing table roots but can't come up with an intuitive example of when someone would want to purposely ignore the "list inconsistency" case.

I think that the compromise of only ignoring deletions at the root and _not_ ignoring deletions at lower levels of the listing represents a fair compromise: we may fail to reliably detect _all_ race conditions but the additional detection still provides some incremental value and I feel like it's unlikely to result in "false positives" which would break real workloads.

@marmbrus @rxin @hvanhovell, do any of you have opinions on this PR's changes? The user-facing migration guide documentation is a bit convoluted here because it's trying to describe what I think is pretty narrow set of circumstances where this change might break workloads.

If we're not comfortable tying this to the `ignoreMissingFiles` configuration, maybe we could add a new configuration specifically for this behavior?

Reply:
One usecase I faced at that time was that files were periodically removed by another script in structured streaming. They wanted to sync the timing somehow with the script and structured streaming. So, some files were missing comparing to the listed in the file index.

I am good as is defaulting to `false` if we're all fine because logically it sounds we should better explicitly set `spark.sql.files.ignoreMissingFiles` in most cases .. but to me I feel like there might be some cases missing and better be stay safe and conservative.

Reply:
Given the we are coming up on a major release I'm inclined to fix this and have one simple flag that controls this behavior (and that defaults to `false`, as most people would expect to be told if there is an underlying inconsistency).

I think it was a mistake to make the changes in #15153, rather than just telling the user to set this flag.